### PR TITLE
Fix Qdrant Race Condition

### PR DESCRIPTION
**Pull request by sourabhdesai**
# Description

The Qdrant Vector DB integration was using `recreate_collection` instead of `create_collection`.

The `recreate_collection` *deletes* the collection if it already exists and creates an empty new collection. This is fine in cases where you're just running a notebook since you just have one thread writing into the vector store. Thats because this integration does its own check to see if the collection already exists before calling `recreate_collection`. However, if you are in a situation with multiple writers each doing parallel writes, this becomes problematic since there could be a race condition where they both check if the collection exist at the same time, but then one writer inserts nodes after that only for them to get wiped out by the other writer that does `recreate_collection`. Probably a highly improbable race condition but easy to handle properly.

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/vector_stores/qdrant.py [hunk 1]
```diff
@@ -308,57 +308,73 @@ def _create_collection(self, collection_name: str, vector_size: int) -> None:
...
-                collection_name=collection_name,
-                vectors_config=rest.VectorParams(
-                    size=vector_size,
-                    distance=rest.Distance.COSINE,
-                ),
+                    ),
+                )
+        except ValueError as exc:
+            if "already exists" not in str(exc):
+                raise exc
```
Comment (sourabhdesai) [inline llama_index/vector_stores/qdrant.py:374]: not personally a fan of handling this way but [the in-memory qdrant client just raises a `ValueError`](https://github.com/qdrant/qdrant-client/blob/efb876fe3915dc5e2855f60a5617e940c84591e5/qdrant_client/local/qdrant_local.py#L618) which is a pretty generic error class. I wanted to have at least some assurance that the `ValueError` was regarding the collection already existing.

File: llama_index/vector_stores/qdrant.py [hunk 2]
```diff
@@ -308,57 +308,73 @@ def _create_collection(self, collection_name: str, vector_size: int) -> None:
...
-            self._client.recreate_collection(
-                collection_name=collection_name,
-                vectors_config=rest.VectorParams(
-                    size=vector_size,
-                    distance=rest.Distance.COSINE,
-                ),
+                    ),
+                )
+        except ValueError as exc:
+            if "already exists" not in str(exc):
```
Comment (hatianzhang) [inline llama_index/vector_stores/qdrant.py:336]: should we just first check whether collection exists instead try/except.
it seems it has get_collection https://github.com/qdrant/qdrant-client/blob/efb876fe3915dc5e2855f60a5617e940c84591e5/qdrant_client/local/qdrant_local.py#L573

Comment (logan-markewich) [reply llama_index/vector_stores/qdrant.py:336]: I think there can still be a race-condition by doing that (two different clients could see that no collection is there at the same time, leading to needing the try/except)

**Comment by logan-markewich** [comment]
lol whoops, pushed a change to the wrong branch. Should be fine though

**Review by hatianzhang — APPROVED** [review]
LGTM

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**